cpp/ql/src/Likely Bugs/Likely Typos/AssignWhereCompareMeant.ql

```diff
@@ -54,14 +54,29 @@ class BooleanControllingAssignmentInExpr extends BooleanControllingAssignment {
   override predicate isWhitelisted() {
     this.getConversion().(ParenthesisExpr).isParenthesised()
     or
-    // whitelist this assignment if all comparison operations in the expression that this
+    // Allow this assignment if all comparison operations in the expression that this
     // assignment is part of, are not parenthesized. In that case it seems like programmer
     // is fine with unparenthesized comparison operands to binary logical operators, and
     // the parenthesis around this assignment was used to call it out as an assignment.
     this.isParenthesised() and
     forex(ComparisonOperation op | op = getComparisonOperand*(this.getParent+()) |
       not op.isParenthesised()
     )
+    or
+    // Match a pattern like:
+    // ```
+    // if((a = b) && use_value(a)) { ... }
+    // ```
+    // where the assignment is meant to update the value of `a` before it's used in some other boolean
+    // subexpression that is guarenteed to be evaluate _after_ the assignment.
+    this.isParenthesised() and
+    exists(LogicalAndExpr parent, Variable var, VariableAccess access |
+      var = this.getLValue().(VariableAccess).getTarget() and
+      access = var.getAnAccess() and
+      not access.isUsedAsLValue() and
+      parent.getRightOperand() = access.getParent*() and
+      parent.getLeftOperand() = this.getParent*()
```

Comment:
I'm worried about the join order that'll be picked here, and I'm not sure an optimal join order even exists. Transitive closures are always risky to join on, and the TC of `getParent` can be very large (and hard to compress) if the optimiser doesn't manage to constrain it. Even the pattern of `access1.getTarget() = access2.getTarget()` can explode if it comes so early in the join order that all combinations of `(access1, access2)` for the same target are computed.

Can you show some tuple counts to convince me that there is no performance risk here?

Often, the solution is to do something more principled than `getParent*` and consider which parent-child relationships we actually want. Conjunctions inside conjunctions only?

Reply:
Yeah, I think conjunctions inside conjunctions could work if the tuple counts end up being a problem. I'll dig into the tuple counts more thoroughly and see if we need to change anything.

Reply:
Here's the RA for the `isWhitelisted` predicate on Linux:
```codeql
>>> Created relation #Expr::Expr::getParent_dispred#bfPlus/2@c09cdw with 9404704 rows.
...
Starting to evaluate predicate AssignWhereCompareMeant::BooleanControllingAssignment::isWhitelisted_dispred#f/1@c448c1
Tuple counts for AssignWhereCompareMeant::BooleanControllingAssignment::isWhitelisted_dispred#f/1@c448c1:
  385      ~0%      {1} r1 = JOIN AssignWhereCompareMeant::BooleanControllingAssignmentInStmt#class#f WITH Expr::Expr::isParenthesised_dispred#f ON FIRST 1 OUTPUT Lhs.0 'this'
  
  66       ~0%      {2} r2 = JOIN AssignWhereCompareMeant::BooleanControllingAssignment::isWhitelisted_dispred#f#shared#3 WITH @cmp_op_expr#f ON FIRST 1 OUTPUT Lhs.1 'this', Lhs.0
  66       ~0%      {2} r3 = r2 AND NOT AssignWhereCompareMeant::BooleanControllingAssignment::isWhitelisted_dispred#f#antijoin_rhs#1(Lhs.0 'this', Lhs.1)
  66       ~0%      {1} r4 = SCAN r3 OUTPUT In.0 'this'
  
  451      ~2%      {1} r5 = r1 UNION r4

  // BooleanControllingAssignmentInExpr::isWhitelisted begins here
  348      ~2%      {2} r6 = JOIN AssignWhereCompareMeant::BooleanControllingAssignmentInExpr#class#f WITH exprconv ON FIRST 1 OUTPUT Rhs.1, Lhs.0 'this'
  7        ~0%      {3} r7 = JOIN r6 WITH Expr::Expr::isParenthesised_dispred#f ON FIRST 1 OUTPUT Lhs.0, 12, Lhs.1 'this'
  7        ~0%      {1} r8 = JOIN r7 WITH exprs ON FIRST 2 OUTPUT Lhs.2 'this'
  
  348      ~0%      {2} r9 = JOIN AssignWhereCompareMeant::BooleanControllingAssignment::isWhitelisted_dispred#f#shared WITH Assignment::Assignment::getLValue_dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.0 'this'
  
  345      ~0%      {2} r10 = JOIN r9 WITH Access::Access::getTarget_dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'this'
  
  4347     ~0%      {2} r11 = JOIN r10 WITH Access::Access::getTarget_dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Lhs.1 'this', Rhs.1
  
  3447     ~0%      {2} r12 = r11 AND NOT Access::VariableAccess::isUsedAsLValue_dispred#f(Lhs.1)
  
  3447     ~0%      {2} r13 = r12 AND NOT Access::VariableAccess::isUsedAsLValue_dispred#f(Lhs.1)
  
  3447     ~0%      {2} r14 = SCAN r13 OUTPUT In.1, In.0 'this'
  // #Expr::Expr::getParent_dispred#bfPlus is magic'd.
  10717    ~8%      {2} r15 = JOIN r14 WITH #Expr::Expr::getParent_dispred#bfPlus ON FIRST 1 OUTPUT Rhs.1, Lhs.1 'this'
  
  14164    ~7%      {2} r16 = r14 UNION r15
  
  193      ~12%     {3} r17 = JOIN r16 WITH Expr::BinaryOperation::getRightOperand_dispred#bf_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, 65, Lhs.1 'this'
  
  193      ~17%     {3} r18 = JOIN r17 WITH exprs ON FIRST 2 OUTPUT Lhs.0, 0, Lhs.2 'this'
  
  12       ~0%      {1} r19 = JOIN r18 WITH Expr::Expr::getChild_dispred#fff ON FIRST 3 OUTPUT Lhs.2 'this'
  12       ~0%      {1} r20 = JOIN r19 WITH AssignWhereCompareMeant::BooleanControllingAssignmentInExpr#class#f ON FIRST 1 OUTPUT Lhs.0 'this'
  12       ~0%      {1} r21 = JOIN r20 WITH Expr::Expr::isParenthesised_dispred#f ON FIRST 1 OUTPUT Lhs.0 'this'
  
  193      ~13%     {2} r22 = JOIN r18 WITH Expr::Expr::getChild_dispred#fff ON FIRST 2 OUTPUT Lhs.2 'this', Rhs.2
  193      ~13%     {2} r23 = JOIN r22 WITH AssignWhereCompareMeant::BooleanControllingAssignmentInExpr#class#f ON FIRST 1 OUTPUT Lhs.0 'this', Lhs.1
  193      ~13%     {2} r24 = JOIN r23 WITH Expr::Expr::isParenthesised_dispred#f ON FIRST 1 OUTPUT Lhs.0 'this', Lhs.1
  // Similar magic here.
  5        ~0%      {1} r25 = JOIN r24 WITH #Expr::Expr::getParent_dispred#bfPlus ON FIRST 2 OUTPUT Lhs.0 'this'
  
  17       ~5%      {1} r26 = r21 UNION r25
  24       ~3%      {1} r27 = r8 UNION r26
  // BooleanControllingAssignmentInExpr::isWhitelisted ends here
  
  475      ~3%      {1} r28 = r5 UNION r27
                    return r28
```

This is magic applied to the transitive closure:
```codeql
magic
m##Expr::Expr::getParent_dispred#bfPlus(/* Expr::Expr */ cached unique entity this)
:-
  project#Assignment::Assignment::getLValue_dispred#ff(this);
  project#Access::Access::getTarget_dispred#ff(this)
.
```

So I think we're saved by magic. It seems to perform well on both Linux and Kamailio as-is, but if you feel more confident with a more principled transitive closure I'll happily do that.

Reply:
Alright, let's try it out.